### ISSUE TITLE
float32 gradients of bfloat16 weights

### ISSUE THREAD

**Issue by martin-marek** [pytorch/pytorch#158709]
### 🚀 The feature, motivation and pitch

I am interested in fine-tuning a large model in a memory-constrained setting. I would like to store model weights in bf16 and compute activations and gradients in fp32. However, I cannot figure out how to compute gradients in a different precision from the weights.

My understanding is that by default, autograd uses the same precision for gradients as the weights. Is there any way to get around this?

**Comment by lokashrinav** [comment]
Mind sharing a bit more about your overall setup? Just wondering if there might be an easier or more efficient way depending on where the memory pressure is really coming from

**Comment by martin-marek** [comment]
Sure. I'm interested in full-fine-tuning Gemma 3 12B on an A100 GPU with 40GB of memory. I am memory constrained but not compute constrained.

Some additional details:
- I can afford to store the model in HBM in bf16 (24GB) but not fp32 (48GB)
- the optimizer step is fused into the backward pass, so I never need to materialize the full model gradients
- I'm using an optimizer with a small state size (SGD / Adafactor), so the optimizer doesn't really take up any memory

Clearly, in this setup, I cannot afford to store the model in fp32 but I would still like to get fp32 gradients. I ran a few experiments with a smaller model in JAX where I compared the use of fp32 vs bf16 gradients (computed on bf16 weights) and I found a significant performance gap between bf16 and fp32 gradients. So I would like to get fp32 gradients even though I can only store the weights in bf16.

**Comment by soulitzer** [comment]
One way to get around this today is with a wrapper tensor subclass, see example here https://github.com/pytorch/ao/blob/262b180cea03d362ba13879eaa38085a73de719b/torchao/float8/float8_training_tensor.py#L242
and wrap your inner tensor with a layer that fakes its dtype. It does require a good amount of boilerplate though, so providing a better way is something we should plan to work on.
